File: little_artist_flutter/lib/services/image_processing_service.dart
import 'dart:typed_data';

import 'package:flutter_image_compress/flutter_image_compress.dart';

/// Compresses and resizes images for storage and thumbnail generation.
class ImageProcessingService {
  static const int _mainMaxDimension = 2048;
  static const int _thumbnailMaxDimension = 512;
  static const int _mainQuality = 70;
  static const int _thumbnailQuality = 60;

  /// Compresses [imageBytes] for storage and generates a thumbnail.
  ///
  /// Returns a record with `imageData` (max 2048px, quality 70) and
  /// `thumbnailData` (max 512px, quality 60).
  Future<({Uint8List imageData, Uint8List thumbnailData})> processForStorage(
    Uint8List imageBytes,
  ) async {
    final imageData = await FlutterImageCompress.compressWithList(
      imageBytes,
      minWidth: _mainMaxDimension,
      minHeight: _mainMaxDimension,
      quality: _mainQuality,
      format: CompressFormat.jpeg,
    );

    final thumbnailData = await FlutterImageCompress.compressWithList(
      imageBytes,
      minWidth: _thumbnailMaxDimension,
      minHeight: _thumbnailMaxDimension,
      quality: _thumbnailQuality,
      format: CompressFormat.jpeg,
    );

    return (
      imageData: Uint8List.fromList(imageData),
      thumbnailData: Uint8List.fromList(thumbnailData),
    );
  }

  /// Generates a thumbnail from [imageBytes] (max 512px, quality 60).
  ///
  /// Returns `null` on failure.
  Future<Uint8List?> generateThumbnail(Uint8List imageBytes) async {
    try {
      final data = await FlutterImageCompress.compressWithList(
        imageBytes,
        minWidth: _thumbnailMaxDimension,
        minHeight: _thumbnailMaxDimension,
        quality: _thumbnailQuality,
        format: CompressFormat.jpeg,
      );
      return Uint8List.fromList(data);
    } catch (_) {
      return null;
    }
  }
}

File: Little Artist/Services/ImageProcessingService.swift
//
//  ImageProcessingService.swift
//  Little Artist
//
//  Resizes and compresses artwork images at ingest time.
//  Produces a two-tier output: a main image (max 2048px, JPEG 0.7)
//  and a thumbnail (max 512px, JPEG 0.6).
//

import UIKit

/// Stateless image processing utility for artwork ingest.
///
/// Call ``processForStorage(image:)`` from camera/scanner captures, or
/// ``processForStorage(data:)`` from photo picker / batch imports.
enum ImageProcessingService {

    // MARK: - Configuration

    /// Maximum longest-edge dimension for the main stored image.
    static let mainMaxDimension: CGFloat = 2048

    /// JPEG compression quality for the main image (0.0–1.0).
    static let mainQuality: CGFloat = 0.7

    /// Maximum longest-edge dimension for the thumbnail.
    static let thumbnailMaxDimension: CGFloat = 512

    /// JPEG compression quality for the thumbnail (0.0–1.0).
    static let thumbnailQuality: CGFloat = 0.6

    // MARK: - Public API

    /// Processes a `UIImage` (from camera or scanner) into compressed main + thumbnail data.
    static func processForStorage(image: UIImage) -> (imageData: Data, thumbnailData: Data) {
        let mainImage = resized(image, maxDimension: mainMaxDimension)
        let thumbImage = resized(image, maxDimension: thumbnailMaxDimension)

        let imageData = mainImage.jpegData(compressionQuality: mainQuality) ?? Data()
        let thumbnailData = thumbImage.jpegData(compressionQuality: thumbnailQuality) ?? Data()

        return (imageData: imageData, thumbnailData: thumbnailData)
    }

    /// Processes raw `Data` (from photo picker) into compressed main + thumbnail data.
    /// Returns `nil` if the data cannot be decoded as an image.
    static func processForStorage(data: Data) -> (imageData: Data, thumbnailData: Data)? {
        guard let image = UIImage(data: data) else { return nil }
        return processForStorage(image: image)
    }

    // MARK: - Private

    /// Resizes an image so its longest edge is at most `maxDimension`.
    /// Returns the original image if it's already smaller.
    private static func resized(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let size = image.size
        let longest = max(size.width, size.height)

        guard longest > maxDimension else { return image }

        let scale = maxDimension / longest
        let newSize = CGSize(
            width: (size.width * scale).rounded(.down),
            height: (size.height * scale).rounded(.down)
        )

        let renderer = UIGraphicsImageRenderer(size: newSize)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
    }
}
